File: frontend/src/app/services/employee.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Globals} from "../global/globals";
import {EmployeeCreateDto, EmployeeDetailDto, EmployeeListDto, EmployeeUpdateDto} from "../dtos/employee";
import {Observable} from "rxjs";

@Injectable({
  providedIn: 'root'
})
export class EmployeeService {

  private employeeBaseUri: string = this.globals.backendUri + '/employee';

  constructor(private httpClient: HttpClient, private globals: Globals) { }

  /**
   * Creates a new employee in the backend.
   *
   * @param employeeCreateDto to create
   * @returns the created employee
   */
  createEmployee(employeeCreateDto: EmployeeCreateDto): Observable<EmployeeListDto> {
    return this.httpClient.post<EmployeeListDto>(this.employeeBaseUri, employeeCreateDto);
  }
  /**
   * Retrieves all employees from the backend.
   *
   * @returns a list of employees
   */
  getAllEmployees(): Observable<EmployeeListDto[]> {
    return this.httpClient.get<EmployeeListDto[]>(this.employeeBaseUri);
  }

  /**
   * Finds an employee by ID.
   *
   * @param id of the employee to retrieve
   * @returns the employee details
   */
  findOneEmployee(id: number): Observable<EmployeeDetailDto> {
    return this.httpClient.get<EmployeeDetailDto>(`${this.employeeBaseUri}/${id}`);
  }

  /**
   * Deletes an employee by ID in the backend.
   *
   * @param id of the employee to delete
   * @returns an observable for the delete operation
   */
  deleteEmployee(id: number): Observable<void> {
    return this.httpClient.delete<void>(`${this.employeeBaseUri}/${id}`);
  }

  /**
   * Updates an existing employee in the backend.
   *
   * @param id the ID of the employee to update
   * @param employeeUpdateDto the data to update
   * @returns the updated employee details
   */
  updateEmployee(id: number, employeeUpdateDto: EmployeeUpdateDto): Observable<EmployeeDetailDto> {
    return this.httpClient.put<EmployeeDetailDto>(`${this.employeeBaseUri}/${id}`, employeeUpdateDto);
  }
}

File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/EmployeeService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.EmployeeCreateDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.EmployeeDetailDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.EmployeeListDto;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.EmployeeUpdateDto;
import at.ac.tuwien.sepr.groupphase.backend.entity.Employee;
import at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException;
import at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException;

import java.util.List;

public interface EmployeeService {

    /**
     * Retrieves all rooms from the repository, ordered by price in ascending order.
     *
     * @return a list of all {@link EmployeeListDto}, sorted by last name in descending order.
     */
    List<EmployeeListDto> findAll();


    /**
     * Creates a new employee based on the provided EmployeeCreateDto.
     *
     * @param employeeCreateDto the DTO containing employee creation details
     * @return the created {@link EmployeeListDto}
     */
    EmployeeListDto create(EmployeeCreateDto employeeCreateDto) throws ValidationException;

    /**
     * Updates an existing employee with the specified ID using the provided update data.
     *
     * <p>Only the fields that are present in the {@code EmployeeUpdateDto} will be updated.
     * Fields that are {@code null} in the DTO will remain unchanged. If a password is provided,
     * it will be securely encoded before being updated.</p>
     *
     * @param id the ID of the employee to update; must not be {@code null}.
     * @param employeeUpdateDto the DTO containing the updated fields for the employee; must not be {@code null}.
     * @return an {@code EmployeeDetailDto} representing the updated employee.
     * @throws IllegalArgumentException if the {@code id} is {@code null}.
     * @throws NotFoundException if no employee with the specified {@code id} exists in the repository.
     */
    EmployeeDetailDto update(Long id, EmployeeUpdateDto employeeUpdateDto) throws NotFoundException, ValidationException;

    /**
     * Finds an employee by their id.
     *
     * @param id of the employee
     * @return the {@link EmployeeDetailDto} of the found @{@link Employee}
     */
    EmployeeDetailDto findOne(Long id);


    /**
     * Deletes an employee by their id.
     *
     * @param id of the employee to be deleted
     */
    void delete(Long id);
}
